src/braket/default_simulator/simulator.py

```diff
@@ -202,32 +212,96 @@ def _validate_and_consolidate_observable_result_types(
+                    measured_qubits,
+                    i,
+                    not_none_observable_list,
+                    qubit_observable_mapping,
+                    none_observable_mapping,
+                    identity_qubits,
+                )
+        for i in sorted(identity_qubits):
+            not_none_observable_list.append(Identity([i]))
+        return not_none_observable_list
+
+    @staticmethod
+    def _assign_observable(
+        whole_observable,
+        hashed_qubit_observable,
+        measured_qubits,
+        target_index,
+        not_none_observable_list,
+        qubit_observable_mapping,
+        none_observable_mapping,
+        identity_qubits,
+    ):
+        # Validate that the same observable is requested for a qubit in the result types
+        hashed_qubit_observable = (
+            hashed_qubit_observable[target_index]
+            if isinstance(hashed_qubit_observable, dict)
+            else hashed_qubit_observable
+        )
+        qubit = measured_qubits[target_index]
+        existing_observable = qubit_observable_mapping.get(qubit)
+        if hashed_qubit_observable == Identity.__name__:
+            if qubit not in qubit_observable_mapping:
+                identity_qubits.add(qubit)
+            # Do nothing if non-identity observable already exists on the qubit
+        else:
+            qubit_observable = (
+                whole_observable[target_index]
+                if isinstance(whole_observable, dict)
+                else whole_observable
+            )
+            if not existing_observable:
+                identity_qubits.discard(qubit)
+                qubit_observable_mapping[qubit] = qubit_observable
+
+                if (
+                    not none_observable_mapping.get(hashed_qubit_observable)
+                    and qubit_observable.measured_qubits.index(qubit) == 0
+                ):
+                    not_none_observable_list.append(qubit_observable)
+            else:
+                DefaultSimulator._validate_same_observable(
+                    existing_observable, qubit_observable, qubit
+                )
+
+    @staticmethod
+    def _validate_same_observable(existing, new, qubit):
+        cls_existing = existing.__class__.__name__
+        cls_new = new.__class__.__name__
+        if cls_existing != cls_new:
+            raise ValueError(
+                f"Qubit {qubit} is already being measured in {cls_existing};"
+                f" cannot measure in {cls_new}."
+            )
+        if cls_existing == Hermitian.__name__:
+            if not np.allclose(existing.matrix, new.matrix):
+                raise ValueError(
+                    f"Qubit {qubit} is already being measured in {existing.matrix};"
+                    f" cannot measure in {new.matrix}."
+                )
+            qubits_existing = existing.measured_qubits
+            qubits_new = new.measured_qubits
+            if (
+                qubits_existing is not None
```

Comment:
Are the first two checks (`qubits_existing is not None and len(qubits_existing) > 1`) necessary? Is it possible that `len(qubits_existing) == 1` but `len(qubits_new) > 1`?

Reply:
The `if not np.allclose(existing.matrix, new.matrix)` check above ensures that `len(qubits_existing) == len(qubits_new)` if neither `qubits_existing` and `qubits_new` are `None` , so `len(qubits_existing) == 1 and len(qubits_new) > 1` is impossible. If `qubits_existing is None`, then `existing` must be a single-qubit observable so don't have to proceed with further checks. At this point, I suppose I can remove `len(qubits_existing) > 1`, since `len(qubits_existing) == 1 and qubits_existing != qubits_new` is impossible.

Reply:
Let's remove `len(qubits_existing) > 1`. It confuses me a bit when the conditions aren't symmetric.